tools/shape_tool/ShapeTool.cpp

```diff
@@ -0,0 +1,305 @@
+#include "../../../../lib/drawing/tile.cpp"
+#include "../../../../lib/tiles/common.cpp"
+
+#include "TileEmbeds.cpp"
+#include "TileWindow.cpp"
+#include "ShapeWindow.cpp"
+#include "TileShapes.cpp"
+
+const string EMBED_spr_icon_shape_tool = SPRITES_BASE + "icon_shape_tool.png";
+
+
+class ShapeTool : Tool
+{
+	
+	private Mouse@ mouse;
+	private TileWindow@ tile_window;
+	private ShapeWindow@ shape_window;
+
+	private bool dragging;
+	private float drag_start_x, drag_start_y;
+	private TileShapes@ custom_shape;
+	
+	ShapeTool(AdvToolScript@ script)
+	{
+		super(script, 'Tiles', 'Shape Tool');
+
+		init_shortcut_key(VK::W, ModifierKey::Alt, -1);
+	}
+
+	void create(ToolGroup@ group) override
+	{
+		set_icon(SPRITE_SET, "icon_shape_tool", 40, 40);
+	}
+	
+	void on_init() override
+	{
+		@mouse = script.mouse;
+
+		UI@ ui = script.ui;
+
+		@tile_window = TileWindow(ui);
+		tile_window.x = 90;
+		tile_window.y = 70;
+		ui.add_child(tile_window);
+		script.window_manager.register_element(tile_window);
+
+		@shape_window = ShapeWindow(ui);
+		shape_window.x = 460;
+		shape_window.y = 70;
+		ui.add_child(shape_window);
+		script.window_manager.register_element(shape_window);
```

Comment:
I don't know how much impact this really has, but I defer creating UI till needed, i.e. in `on_select_impl`.